File: typescript/linked-list/linked-list.ts
export class LinkedList<TElement> {
  private head: Node<TElement> | Empty = null
  private last: Node<TElement> | Empty = null
  private length: number = 0

  push = (element: TElement): void => {
    const node = {
      value: element,
      next: null,
      prev: null,
    }

    if (this.isEmpty()) {
      this.head = node
      this.last = node
      this.length += 1
      return
    }

    if (this.last) {
      const next = {
        ...node,
        prev: this.last,
      }

      this.last.next = next 
      this.last = next
      this.length += 1
    }
  }

  isEmpty = (): boolean => this.length === 0

  pop = (): (TElement | Empty) => {
    if (this.last) {
      const node = this.last
      this.last = node.prev
      this.length -= 1
      return node.value
    }

    return null
  }

  shift = (): (TElement | Empty) => {
    if (this.head) {
      const node = this.head
      this.head = node.next
      this.length -= 1
      return node.value
    }
    
    return null
  }

  public unshift(element: TElement) {
    const node = {
      value: element,
      next: null,
      prev: null,
    }

    if (this.isEmpty()) {
      this.head = node
      this.last = node

      this.length += 1
      return
    }
    
    if (this.head) {
      const prev = {
        ...node,
        next: this.head,
      }

      this.head.prev = prev
      this.head = prev
      this.length += 1
    }
  }

  delete = (element: unknown): void => {
    if (this.isEmpty()) return

    if (this.head && this.head.value === element) {
      this.head = this.head.next
      this.length -= 1
      return
    }

    let node = this.head
    while (node) {
      if (node.next && node.next.value === element) {
        node.next = node.next.next
        this.length -= 1
        return
      }

      node = node.next
    }
  }

  count = (): number => {
    return this.length
  }
}

type Empty = null

type Node<TElement> = {
  value: TElement,
  next: Node<TElement> | Empty,
  prev: Node<TElement> | Empty,
}

File: solutions/zig/linked-list/1/linked_list.zig
const std = @import("std");

pub fn main(_: std.process.Init) !void {
    const a: ?usize = 1;
    var x: usize = 0;

    const b = a orelse {
        x = 1;
        return;
    };

    std.debug.print("b {any}", .{b});
}

pub fn LinkedList(comptime T: type) type {
    return struct {
        // Please implement the doubly linked `Node` (replacing each `void`).
        pub const Node = struct {
            prev: ?*Node = null,
            next: ?*Node = null,
            data: T = 0,
        };

        // Please implement the fields of the linked list (replacing each `void`).
        first: ?*Node = null,
        last: ?*Node = null,
        len: usize = 0,

        // Please implement the below methods.
        // You need to add the parameters to each method.

        pub fn push(self: *LinkedList(T), node: *Node) void {
            const last = self.last orelse {
                self.first = node;
                self.last = node;
                self.len += 1;
                return;
            };

            last.next = node;
            node.prev = last;
            self.last = node;
            self.len += 1;
        }

        pub fn pop(self: *LinkedList(T)) ?*Node {
            const last = self.last;
            if (last == null) return null;

            self.last = last.?.prev;
            self.len -= 1;

            return last;
        }

        pub fn shift(self: *LinkedList(T)) ?*Node {
            const first = self.first orelse return null;
            const next = first.next orelse {
                self.first = null;
                self.last = null;
                self.len -= 1;
                return first;
            };

            self.first = next;
            self.first.?.prev = null;
            self.len -= 1;

            return first;
        }

        pub fn unshift(self: *LinkedList(T), node: *Node) void {
            const first = self.first orelse {
                self.first = node;
                self.last = node;
                self.len += 1;
                return;
            };

            self.first = node;
            self.first.?.next = first;
            first.prev = self.first;

            self.len += 1;
        }

        pub fn delete(self: *LinkedList(T), node: *Node) void {
            const prev = node.prev orelse {
                if (self.first == node) _ = self.shift();
                return;
            };

            const next = node.next orelse {
                if (self.last == node) _ = self.pop();
                return;
            };

            prev.next = next;
            next.prev = prev;

            self.len -= 1;
        }
    };
}
